packages/components/MenuButton/macos/MenuButton.swift

```diff
@@ -0,0 +1,111 @@
+//
+//  Copyright (c) Microsoft Corporation. All rights reserved.
+//  Licensed under the MIT License.
+//
+
+import AppKit
+
+@objc(MSFMenuButton)
+open class MenuButton: NSPopUpButton {
+
+  public override init(frame buttonFrame: NSRect, pullsDown flag: Bool) {
+
+    super.init(frame: buttonFrame, pullsDown: flag)
+
+    imagePosition = .imageLeading
+    bezelStyle = .recessed
+
+    if #available(OSX 11.0, *) {
+      controlSize = .large
+    }
+
+    guard let dropDownCell = cell as? NSPopUpButtonCell else {
+      preconditionFailure()
+    }
+    dropDownCell.imagePosition = .imageLeading
+    dropDownCell.arrowPosition = .arrowAtBottom
+  }
+
+  @available(*, unavailable)
+  required public init?(coder decoder: NSCoder) {
+    preconditionFailure()
+  }
+
+  @objc public convenience init() {
+    self.init(frame: .zero, pullsDown: true)
+  }
+
+  @objc public var onPress: RCTBubblingEventBlock?
+
+  open override var menu: NSMenu? {
+    didSet {
+      updateMenu()
+    }
+  }
+
+  open override var image: NSImage? {
+    get {
+      return menuButtonImage
+    }
+    set {
+      // We must set the image on the dropdown cell rather than the button,
+      // If we set the image on Button itself, no image is displayed.
+      menuButtonImage = newValue
+      updateDropDownCell()
+    }
+  }
+
+  open override var title: String {
+    didSet {
+      updateDropDownCell()
+    }
+  }
+
+  // MARK: - Private Methods
+
+  private func updateMenu() {
+    guard let menu = menu else {
+      return
+    }
+
+    // Insert an initial empty item into index 0, since index 0 is never displayed
```

Comment:
Is this documented somewhere? I'm probably missing context but it sounds fishy that an item in the menu wouldn't ever be displayed...

Reply:
Not really. I believe it is because we are using a nonstandard (but accepted) to set the title/image of this MenuButton.

By default, NSPopupButton will either display the first menu item or the currently selected menu item (depending on if the boolean `pullsDown` is set) to display in the "button" part. We don't want to use a menu item from the menu, we want to have a custom title and image that is NOT in the menu. To achieve this, apple recommends setting the NSPopupButton's cell's `usesItemFromMenu` property to false, and then providing that cell a custom menu item to render instead. This is what I am doing from lines 87 to 93. My guess is a side effect is that first menu item that was supposed to be used to render the title/image of the button is now never shown. 

This info comes from this sample code, which is pretty old (pre ARC) but is the only apple official sample code I found https://developer.apple.com/library/archive/samplecode/MenuMadness/Introduction/Intro.html#//apple_ref/doc/uid/DTS40008870

Reply:
for my own knowledge, does VO properly ignores the empty one ?

Reply:
https://user-images.githubusercontent.com/6722175/128072198-0ec89a1e-7d89-4b4c-a790-d695711f2f61.mov

Looks like it.